### ISSUE TITLE
Sometimes card border shown, sometimes it doesn't

### ISSUE THREAD

**Issue by smenzer** [ofekashery/vertical-stack-in-card#150]
This is a bit of a weird one.... I'm using a vertical stack in card with two horizontal stacks inside. Sometimes when I load the dashboard (could be on mobile app, my laptop via the web, etc) I see borders between the cards in the horizontal stack. If I refresh the page, the borders go away. Sometimes I open it back up and the borders are there again. I don't want the borders to appear, but I can't figure out how to consistently keep them away.

Here's a screenshot of me opening the same dashboard twice in a row, once where the borders appeared and once where it didn't.

![image](https://user-images.githubusercontent.com/624392/224556845-a95096de-e940-446c-966a-55a88705f4cb.png)

As you can see, on the right side where it appears the way I want, there are styles added directly to the `<mushroom-template-card>` and `<ha-card>` to remove the border. I'm not using card-mod or anything like that, so I don't know why these appear sometimes and not others. Below is the yaml code for the Living Room card (w/some details removed to make it easier to read):

```yaml
type: custom:vertical-stack-in-card
cards:
  - type: horizontal-stack
    cards:
      - type: custom:mushroom-title-card
        title: Living Room
  - type: horizontal-stack
    cards:
      - type: custom:mushroom-template-card
        primary: Lights
        icon: mdi:lightbulb-group-off
        entity: light.group_upstairs_lights_living_area
        tap_action: ....
        layout: vertical
      - type: custom:mushroom-climate-card
        entity: climate.upstairs_hallway_thermostat
        name: Climate
        layout: vertical
      - type: custom:mushroom-template-card
        primary: TV Outlet
        icon: mdi:power-plug
        entity: switch.living_room_smart_outlet_tv_bottom
        icon_color: green
        layout: vertical
        fill_container: true
      - type: custom:mushroom-template-card
        primary: TV
        icon: si:messenger
        entity: select.living_room_harmony_hub_activities
        icon_color: purple
        layout: vertical
```

I've been reading that I can use card-mod to remove the borders...which I might do. Bit I wanted to share since this seems like a bug, and I'd rather avoid adding a bunch of card-mods if I can avoid it.

Thanks!

**Comment by pbouna321** [comment]
Same thing happening to me here.

**Comment by pbouna321** [comment]
> Same thing happening to me here.

So I just noticed this is only happening on the very first page of my dashboard. The other pages render just fine. Very weird.

**Comment by Aasikki** [comment]
I have this issue as well.

**Comment by colossatr0n** [comment]
I'm having this issue as well, but for me, it's only appearing on my mobile device while using the HA app for iOS.

**Comment by pbouna321** [comment]
> 

Yeah I was deleting spaces at the end of the yaml, then saving because it drives me nuts lol.

**Comment by colossatr0n** [comment]
_If anyone is using  [card_mod HACs integration](https://github.com/thomasloven/lovelace-card-mod) inside the `vertical-stack-in-card`, the following might help._

Since I'm seeing this issue in the iOS app, I figured the issue would show up in Safari, and it did. After inspecting the DOM, I found out that adding `border: none` to `ha-card` in the `card_mod` yaml fixed the issue in both Safari and the iOS app.

Here's a simplified version of my card:
```yaml
type: custom:vertical-stack-in-card
cards:
  - type: entities
    entities:
      - entity: timer.timer2
    card_mod:
      style: # vvv This fixed the border issue for this card
        .: |
          ha-card { border: none; } 
# ...
```
Not sure why Safari/iOS needs this fix for it to render properly when the Brave browser works just fine without it.

**Comment by smenzer** [comment]
> _If anyone is using [card_mod HACs integration](https://github.com/thomasloven/lovelace-card-mod) inside the `vertical-stack-in-card`, the following might help._
> 
> Since I'm seeing this issue in the iOS app, I figured the issue would show up in Safari, and it did. After inspecting the DOM, I found out that adding `border: none` to `ha-card` in the `card_mod` yaml fixed the issue in both Safari and the iOS app.

I tried this and it no luck for me when I added it to the 2nd horizontal stack in my original post.

```yaml
type: custom:vertical-stack-in-card
cards:
  - type: horizontal-stack
    cards:
      - type: custom:mushroom-title-card
        title: Living Room
  - type: horizontal-stack
    card_mod:
      style:
        .: |
          ha-card { border: none; } 
    cards:
      - type: custom:mushroom-template-card
        primary: Lights
        icon: mdi:lightbulb-group-off
        entity: light.group_upstairs_lights_living_area
        tap_action: ....
        layout: vertical
...
```

It still works fine on my chrome browser / desktop, but in the app on my iphone it doesn't help.

**Comment by pbouna321** [comment]
Seems like a lot of us are having this same issue. I wonder what happened with the code.

**Comment by colossatr0n** [comment]
@smenzer Do you have access to a Safari browser? If you do, you'll be able to figure out the exact element that needs `border: none`.

**Comment by smenzer** [comment]
I do, I’ll play with it tonight to see what’s going on.

**Comment by smenzer** [comment]
So I tried adjusting the `card_mod` and that hasn't changed anything. It almost is as if the card-mod code is not running on the page when the borders are showing up...maybe there's a timing issue in loading the code, or a conflict, or cacheing issues? It's less about the selector since it is working sometimes, it's just missing code when it's broken.

Note - I also get this incorrect behavior in Chrome, so this isn't just a Safari issue.

**Comment by pbouna321** [comment]
> So I tried adjusting the `card_mod` and that hasn't changed anything. It almost is as if the card-mod code is not running on the page when the borders are showing up...maybe there's a timing issue in loading the code, or a conflict, or cacheing issues? It's less about the selector since it is working sometimes, it's just missing code when it's broken.
> 
> Note - I also get this incorrect behavior in Chrome, so this isn't just a Safari issue.

It's for sure not just on Safari, It is happening on my Android phone and Brave browser. I noticed that the code is sometimes there and sometimes not there like you said too.

**Comment by pbouna321** [comment]
> I do, I’ll play with it tonight to see what’s going on.

Any luck?

**Comment by smenzer** [comment]
the element that the border is set on seems to be the `<ha-card>` using `:host` to apply the border.

![image](https://user-images.githubusercontent.com/624392/236317386-177a272d-1ba8-4262-80c6-e25824e51b35.png)


applying the `card_mod` to the individual cards in the horizontal stacks seems to have done the trick, though that's more tedious and annoying to change later, so i'm still trying to find a way to set that at the higher level. 

what's so strange, though, is that it works sometimes at the horizontal stack level:

<img width="374" alt="image" src="https://user-images.githubusercontent.com/624392/236317885-efd65205-bf32-4078-af08-a4122178cc2a.png">

all i did was refresh the page and sometimes it works - in this instance above, you can see that the `<ha-card>` properly had the styles applied to it.

so, in summary - applying the `card_mod` to every card in the stack works (every time so far). applying it at the horizontal-stack level works most of the time, but not every time.

**Comment by Kanecaine** [comment]
> applying the `card_mod` to the individual cards in the horizontal stacks seems to have done the trick, though that's more tedious and annoying to change later, so i'm still trying to find a way to set that at the higher level.

I just might found the solution for this: use card mod themes.

```yaml
Stacked_Cards:
  card-mod-theme: Stacked_Cards
  card-mod-card: |
    ha-card.type-custom-vertical-stack-in-card {
      border: none;
    }
```

More info on how to use it: 

https://github.com/thomasloven/lovelace-card-mod/wiki/Card-mod-Themes

I recently added this in a custom theme file and all borders are gone without having to add this style to every card. If you are already using a theme you can just simply add this to it.

**Comment by Kanecaine** [comment]
Hello again,
i gained further experience with my solution posted before. It's just not ideal because every card of this type will loose its border. So i tweaked it to this:

```yaml
Stacked_Cards:
  card-mod-theme: Stacked_Cards
  card-mod-card: |
    /* container elements (consume the style) */
    ha-card.type-tile {
      border-width: var(--stacked-border-width, 1px);
    }

    /* container (define the style) */
    ha-card.type-custom-vertical-stack-in-card {
      --stacked-border-width: 0px;
    }

    /* extra classes - use in card yaml:
        card_mod:
          class: no-border no-bg
    */
    ha-card.no-border {
      box-shadow: none;
      border: 0px;
      border-radius: 0;
    }
    ha-card.no-bg {
      background: transparent;
    }
```

All elements (e.g. tile cards) in a container like the vertical stack card are getting a style which uses a css variable. The trick is to set the variable only for the container. So only tile cards within a vertical stack card are styled and have no borders. All other tile cards, somewhere in the view outside our container, are kepping there borders.

So you have to define this for every "container" you are using. No matter if this is a horizontal stack or Thomas Loven's layout card. And you have to define this for every card-type you are using ins such containers. In my example every tile card in vertical stacks is covered. 

At last i defined some extra classes which can be assigned for single use e.g. when you are using a card only once inside a container and you do not want to write a rule for that card-type. You can combine these classes and decide per card if it should have only no border or no background too. Or you define what else you need.

In fact this makes the `vertical-stack-in-card` quite obsolete because its main feature for removing the borders can now be done with every "container card". I recommend using [layout-card](https://github.com/thomasloven/lovelace-layout-card) because of its flexible layout options.

For further inspiration on how to use card mod themes have a look at the corresponding thread, especially [this post](https://community.home-assistant.io/t/card-mod-super-charge-your-themes/212176/1323)

**Comment by natejhowe** [comment]
+1. This is still an issue in the latest version. It seems to specifically be a problem with the vertical-stack element.

**Comment by schrammtastisch** [comment]
> _If anyone is using [card_mod HACs integration](https://github.com/thomasloven/lovelace-card-mod) inside the `vertical-stack-in-card`, the following might help._
> 
> Since I'm seeing this issue in the iOS app, I figured the issue would show up in Safari, and it did. After inspecting the DOM, I found out that adding `border: none` to `ha-card` in the `card_mod` yaml fixed the issue in both Safari and the iOS app.
> 
> Here's a simplified version of my card:
> 
> ```yaml
> type: custom:vertical-stack-in-card
> cards:
>   - type: entities
>     entities:
>       - entity: timer.timer2
>     card_mod:
>       style: # vvv This fixed the border issue for this card
>         .: |
>           ha-card { border: none; } 
> # ...
> ```
> 
> Not sure why Safari/iOS needs this fix for it to render properly when the Brave browser works just fine without it.

Thanks!! This did it for me where borders would show up after every refresh on both Windows and Android :D

**Comment by 931122** [comment]
this is mycode
  ` card_mod:
      style: |
        ha-card {
        box-shadow: none;
        background: none;
        border-bottom-right-radius: 0px;
        border-bottom-left-radius: 0px;
        border-top-right-radius: 0px;
        border-top-left-radius: 0px;
        }`

**Comment by kpagnat** [comment]
Hi all,

Had the same problem but I wasn't willing to use card_mod.
I just take a look into the "vertical-stack-in-card.js" and found out that I could add just 1 line of code to make the border disapear:
You just have to replace the _styleCard(element) method with this one:

```
_styleCard(element) {
    const config = this._config;
    if (element.shadowRoot) {
      if (element.shadowRoot.querySelector('ha-card')) {
        let ele = element.shadowRoot.querySelector('ha-card');
        ele.style.boxShadow = 'none';
        ele.style.borderRadius = '0';
        ele.style.setProperty('border', '0', 'important');
        if ('styles' in config) {
          Object.entries(config.styles).forEach(([key, value]) =>
            ele.style.setProperty(key, value)
          );
        }
      } else {
        let searchEles = element.shadowRoot.getElementById('root');
        if (!searchEles) {
          searchEles = element.shadowRoot.getElementById('card');
        }
        if (!searchEles) return;
        searchEles = searchEles.childNodes;
        for (let i = 0; i < searchEles.length; i++) {
          if (searchEles[i].style) {
            searchEles[i].style.margin = '0px';
          }
          this._styleCard(searchEles[i]);
        }
      }
    } else {
      if (
        typeof element.querySelector === 'function' &&
        element.querySelector('ha-card')
      ) {
        let ele = element.querySelector('ha-card');
        ele.style.boxShadow = 'none';
        ele.style.borderRadius = '0';
        ele.style.setProperty('border', '0', 'important');
        if ('styles' in config) {
          Object.entries(config.styles).forEach(([key, value]) =>
            ele.style.setProperty(key, value)
          );
        }
      }
      let searchEles = element.childNodes;
      for (let i = 0; i < searchEles.length; i++) {
        console.log(searchEles[i]);
        if (searchEles[i] && searchEles[i].style) {
          searchEles[i].style.margin = '0px';
        }
        this._styleCard(searchEles[i]);
      }
    }
  }
```

For those interested, the lines I added are: `ele.style.setProperty('border', '0', 'important');` twice.
It sets the border of the ha-card to 0 and force it with the "!important" css attribute.

**IMPORTANT** : you have to remove the file ".js.gz" file next to it to prevent HA from using cache.

Maybe this could be added to the original code.
Have a nice day.

**Comment by ofekashery** [comment]
Thanks for your suggestion @kpagnat! Can you please share your card config so I can reproduce this issue?

**Comment by kpagnat** [comment]
Hello @ofekashery 

My card config is as follow:
```
type: custom:vertical-stack-in-card
cards:
  - type: custom:mushroom-template-card
    primary: ""
    secondary: ""
    icon: mdi:human-male-female
    icon_color: "#624a8e"
    layout: vertical
  - type: custom:bignumber-card
    entity: sensor.TEMPERATURE1
    scale: 16px
    round: 1
    color: "#624a8e"
    positions:
      icon: "off"
      indicator: "off"
      name: "off"
      title: "off"
      value: inside
    noneString: Off.
  - type: custom:bignumber-card
    entity: sensor.HUMIDITY1
    scale: 12px
    round: 1
    color: "#777"
    positions:
      icon: "off"
      indicator: "off"
      name: "off"
      title: "off"
      value: inside
    noneString: Off.

```

**Comment by ahmedelhaw** [comment]
> _If anyone is using [card_mod HACs integration](https://github.com/thomasloven/lovelace-card-mod) inside the `vertical-stack-in-card`, the following might help._
> 
> Since I'm seeing this issue in the iOS app, I figured the issue would show up in Safari, and it did. After inspecting the DOM, I found out that adding `border: none` to `ha-card` in the `card_mod` yaml fixed the issue in both Safari and the iOS app.
> 
> Here's a simplified version of my card:
> 
> type: custom:vertical-stack-in-card
> cards:
>   - type: entities
>     entities:
>       - entity: timer.timer2
>     card_mod:
>       style: # vvv This fixed the border issue for this card
>         .: |
>           ha-card { border: none; } 
> # ...
> Not sure why Safari/iOS needs this fix for it to render properly when the Brave browser works just fine without it.

` border: none;` solved issue on iPad and Chrome for me! Thanks

**Comment by Ali229** [comment]
```
card_mod:
  style: |
    ha-card {
      border: none;
    }
```
Used to work but it doesn't anymore. Besides, Isn't the whole point of vertical-stack-in-card to stack multiple cards without border? Why would we use card_mod for this. Here is my current setup and result:


```
type: custom:vertical-stack-in-card
cards:
  - type: alarm-panel
    entity: alarm_control_panel.home_alarm
    states:
      - arm_away
      - arm_home
    name: Ring Alarm
  - type: entities
    entities:
      - entity: switch.home_siren
        name: Activate Siren
      - entity: binary_sensor.motion_detector
        name: Living Room Motion
      - entity: binary_sensor.door_motion
        icon: mdi:door
```

![Image](https://github.com/user-attachments/assets/00f34ba2-214a-4269-9e26-b429937e27ea)

Also note that this border doesn't show up when using visual editor, just shows up on dashboard, so definitely a bug, let me know if you need help debugging this further.

**Comment by MathisP75** [comment]
This is still an issue. Sometimes, the borders are rounded, and sometimes they're not, it's very random.
![Screenshot_20250416-103159.png](https://github.com/user-attachments/assets/c282e6df-fe9c-4add-a865-23767ec0b66e)

![Screenshot_20250416-103209.png](https://github.com/user-attachments/assets/312d983f-856c-4dc5-948d-3047eaa873e6)

**Comment by ofekashery** [comment]
This issue has been fixed in v1.1.0. I'll go ahead and close this now, but feel free to reopen if you're still experiencing it after updating.
